pkg/resources/common/common.go

```diff
@@ -14,3 +19,50 @@ func ConvertAuthnExtras(extra map[string]authnv1.ExtraValue) map[string]authzv1.
 	}
 	return result
 }
+
+// ValidateLabel checks if a user is removing or modifying a label. If the label is newly added, return false.
+func IsModifyingLabel(oldLabels, newLabels map[string]string, label string) bool {
+	var oldValue, newValue string
+	var oldLabelExists, newLabelExists bool
+	if oldLabels == nil {
+		oldLabelExists = false
+	} else {
+		oldValue, oldLabelExists = oldLabels[label]
+	}
+	if newLabels == nil {
+		newLabelExists = false
+	} else {
+		newValue, newLabelExists = newLabels[label]
+	}
+
+	if !oldLabelExists {
+		return false
+	}
+	if oldLabelExists && !newLabelExists {
+		return true
+	}
+	if oldValue != newValue {
+		return true
+	}
+
+	return false
+}
+
+// NewVerbChecker returns a function that checks if a user has the specified verb
```

Comment:
The code becomes more complex if we want to 
1. Return all possible validation error instead of just the first one encountered
2. Want to do an escalation check whenever we find an error
3. Only do an escalation once

Caching whether we've done the escalation check and its result makes it simpler.

Though I think we'd want it to be clear that it's a "cached result" type of thing. `NewVerbChecker` doesn't really describes that.

I think either renaming the function (and changing its logic a bit) would make it clearer.

Something like:

```go
type cachedEscalationChecker struct {
  checked bool
  result bool
}

func NewCachedEscalationChecker() *cachedEscalationChecker {
    return cacheEscalationChecker{}
}

func (c *cachedEscalationChecker) CanEscalate(...) (bool, error) {
    if c.checked {
        return c.result // return result, not false
    }
    // actually do the escalation check
    c.checked = true
    c.result = <check>
    return c.result
}
```

Then using it is pretty much unchanged from how the PR uses it.

(Also, naming things is hard, huh...)

>  I think this additional abstraction is adding complexity in an are where it can be dangerous

I'd say no abstraction here was more complex imo and more difficult to reason about so more dangerous.

What's your thought on the proposed changes above, would that still be too abstracted in your opinion?

Reply:
> The code becomes more complex if we want to

The first requirement is more fungible than the others, but sure, there is definitely some extra complexity introduced.

> Caching whether we've done the escalation check and its result makes it simpler

The issue here is that this code doesn't effectively handle that. 

> Do you think the proposed changes above would make it less complex? Or are you opposed in a caching behavior for this area of the code?

I don't think that the proposed changes reduce complexity, particularly with the issues that it introduces. I'm not opposed to making sure that the escalate verb is checked for at most one time, but I don't think this abstraction solves that problem in a secure fashion.

I think that maintaining the state in the Admit function with two bools (and possibly using a side function to do the debouncing/caching) is a better option. However, I'm open to discussion.

Reply:
> using a side function to do the debouncing/caching

Can you expand on this? Sounds like something similar to above but more "local" to the code that needs it.

Reply:
Sure, I'm thinking something like:

```golang

func (a *admitter) Admit(request *admission.Request) (*admissionv1.AdmissionResponse, error) {
// some other code before
var hasEscalate, escalateChecked bool
err = auth.ConfirmNoEscalation(request, clusterRules, "", a.grbResolver)
if err != nil {
	hasEscalate, escErr := a.hasEscalate(hasEscalate, escalateChecked, request, newGR.Name)
	if err != nil {
		return nil, escErr
	}
	escalateChecked = true
	if hasEscalate {
		return admission.ResponseAllowed(), nil
	}
	returnError = errors.Join(returnError, err)
}

// some other code after
}

func (a *admitter) hasEscalate(alreadyChecked bool, prevAnswer bool, request *admission.Request, grName string) (bool, error) {
	if alreadyChecked {
		return prevAnswer, nil
	}
	hasVerb, err := auth.RequestUserHasVerb(request, gvr, a.sar, escalateVerb, grName, "")
	if err != nil {
		return false, fmt.Errorf("unable to check if user has escalate: %w", err)
	}
	return hasVerb, nil
}

```

This has the following benefits over the current approach:
- The specific requirements of the GR/GRB validator (i.e. escalate is only checked once, but we only care about that one resource for any given request) are kept internal to that validator
- The state of what was called and when is kept inside the admit function, which is directing the flow and maintaining the state for this call (rather than in another call elsewhere

This is just my initial answer, if you see other ways to structure the side functions I'm also open to that. My key concern here is making sure that what we do here